docs/plugins.md

```diff
@@ -0,0 +1,138 @@
+that matches Github's requirements for `username/org` and `repository name`. This requirement is in place to help allow multiple creators
+of the same plugin type to exists such as `<org1>/nginx` and `<org2>/nginx`. These names could be based of the repo name 
+such as `argoproj-labs/sample-rollouts-metric-plugin` but it is not a requirement. 
+
+There will also be a standard for naming repositories under argoproj-labs in the form of `rollouts-<type>-<tool>-plugin`
+where `<type>` is say `metric`, or `trafficrouter` and `<tool>` is the software the plugin is for say nginx.
+
+## Plugin Name
+
+So now that we have an idea on plugin naming and repository standards let's pick a name to use for the rest of this 
+documentation and call our plugin `argoproj-labs/nginx`.
+
+This name will be used in a few different spots the first is the config map that your plugin users will need to configure.
+It looks like this below.
+
+```yaml
+kind: ConfigMap
+metadata:
+  name: argo-rollouts-config
+data:
+  plugins: |-
+    metrics:
+    - plugin: "argoproj-labs/metrics"
+      pluginLocation: "file:///Users/zaller/Development/argo-rollouts/metric-plugin"
+    trafficrouters:
+    - plugin: "argoproj-labs/nginx"
+      pluginLocation: "file:///tmp/argo-rollouts/traffic-plugin"
+```
+
+As you can see there is a field called `plugin:` under both `metrics` or `trafficrouters` this is the first place where your
+end users will need to configure the name of the plugin. The second location is either in the rollout object or the analysis
+template which you can see the examples below.
+
+#### AnalysisTemplate Example
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: AnalysisTemplate
+metadata:
+  name: success-rate
+spec:
+  metrics:
+    - name: success-rate
+      ...
+      provider:
+        plugin:
+          argoproj-labs/metrics:
+            address: http://prometheus.local
+```
+
+#### Traffic Router Example
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Rollout
+metadata:
+  name: example-plugin-ro
+spec:
+  strategy:
+    canary:
+      canaryService: example-plugin-ro-canary-analysis
+      stableService: example-plugin-ro-stable-analysis
+      trafficRouting:
+        plugin:
+          argoproj-labs/nginx:
+            stableIngress: canary-demo
+```
+
+You can see that we use the plugin name under `spec.metrics[].provider.plugin` for analysis template and `spec.strategy.canary.trafficRouting.plugin`
+for traffic routers. You as a plugin author can then put any configuration you need under that object and you will be able to
+look up that config in your plugin via the plugin name key.
+
+## Plugin Interfaces
+
+Argo Rollouts currently supports two plugin systems as a plugin author your end goal is to implement these interfaces as
+a hashicorp go-plugin. The two interfaces are `MetricsPlugin` and `TrafficRouterPlugin` for each of the respective plugins:
+
+```go
+type MetricsPlugin interface {
+	NewMetricsPlugin(v1alpha1.Metric) types.RpcError
```

Comment:
Any reason this can't just be `New`?

Reply:
I don't think there is a reason not to I just don't really see that in golang convention you generally see NewSomething https://go.dev/doc/effective_go#composite_literals

Reply:
and all other New functions within rollouts and std lib etc are NewSomething

Reply:
I think that convention typically applies to constructors, not initializers. i.e. return type `MetricsPlugin`.

Maybe `Init`?